lib/upipe-hbrmt/upipe_pack10bit.c

```diff
@@ -0,0 +1,389 @@
+ * @param upump_p reference to pump that generated the buffer
+ */
+static void upipe_pack10bit_input(struct upipe *upipe, struct uref *uref,
+                            struct upump **upump_p)
+{
+    if (!upipe_pack10bit_check_input(upipe)) {
+        upipe_pack10bit_hold_input(upipe, uref);
+        upipe_pack10bit_block_input(upipe, upump_p);
+    } else if (!upipe_pack10bit_handle(upipe, uref, upump_p)) {
+        upipe_pack10bit_hold_input(upipe, uref);
+        upipe_pack10bit_block_input(upipe, upump_p);
+        /* Increment upipe refcount to avoid disappearing before all packets
+         * have been sent. */
+        upipe_use(upipe);
+    }
+}
+
+/** @internal @This receives a provided ubuf manager.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_format amended flow format
+ * @return an error code
+ */
+static int upipe_pack10bit_check(struct upipe *upipe, struct uref *flow_format)
+{
+    struct upipe_pack10bit *upipe_pack10bit = upipe_pack10bit_from_upipe(upipe);
+    if (flow_format != NULL)
+        upipe_pack10bit_store_flow_def(upipe, flow_format);
+
+    if (upipe_pack10bit->flow_def == NULL)
+        return UBASE_ERR_NONE;
+
+    bool was_buffered = !upipe_pack10bit_check_input(upipe);
+    upipe_pack10bit_output_input(upipe);
+    upipe_pack10bit_unblock_input(upipe);
+    if (was_buffered && upipe_pack10bit_check_input(upipe)) {
+        /* All packets have been output, release again the pipe that has been
+         * used in @ref upipe_pack10bit_input. */
+        upipe_release(upipe);
+    }
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This sets the input flow definition.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_def flow definition packet
+ * @return an error code
+ */
+static int upipe_pack10bit_set_flow_def(struct upipe *upipe, struct uref *flow_def)
+{
+    if (flow_def == NULL)
+        return UBASE_ERR_INVALID;
+
+    UBASE_RETURN(uref_flow_match_def(flow_def, "block."))
+
+    uint64_t align;
+    UBASE_RETURN(uref_block_flow_get_align(flow_def, &align))
+    if (!align || align % 16)
+        return UBASE_ERR_INVALID;
+
+    struct uref *flow_def_dup = uref_dup(flow_def);
+
+    if (flow_def_dup == NULL)
+        return UBASE_ERR_ALLOC;
+
+    uref_flow_set_def(flow_def_dup, "block.");
+
+    upipe_input(upipe, flow_def_dup, NULL);
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This processes control commands on a file source pipe, and
+ * checks the status of the pipe afterwards.
+ *
+ * @param upipe description structure of the pipe
+ * @param command type of command to process
+ * @param args arguments of the command
+ * @return an error code
+ */
+static int upipe_pack10bit_control(struct upipe *upipe, int command, va_list args)
+{
+    switch (command) {
+        case UPIPE_REGISTER_REQUEST: {
+            struct urequest *request = va_arg(args, struct urequest *);
+            return upipe_pack10bit_alloc_output_proxy(upipe, request);
```

Comment:
You shouldn't forward the ubuf_mgr and flow_format requests, as you are reallocating the ubufs.